File: September-LeetCoding-Challenge/11-Maximum-Product-Subarray/Maximum-Product-Subarray.cs
public class Solution {
    public int MaxProduct(int[] nums)
    {
        int s = nums[0];
        int r = nums[0];
        
        for(int i=1, max = s, min = s; i<nums.Length; i++){
            int temp = max;
            max = Math.Max(Math.Max(nums[i] * max, nums[i] * min), nums[i]);
            min = Math.Min(Math.Min(nums[i] * temp, nums[i] * min), nums[i]);
            
            r = Math.Max(max, r);
        }
        
        return r;
    }
}

File: September-LeetCoding-Challenge/11-Maximum-Product-Subarray/Maximum-Product-Subarray.rb
def max_product(nums)
    
    min = []
    min[0] = nums[0]
    max = []
    max[0] = nums[0]
    
    nums.each_with_index do |val, n|

        next if n == 0
        min[n] = [val, val * min[n-1], val * max[n-1]].min
        max[n] = [val, val * min[n-1], val * max[n-1]].max
        
    end
    max.max
end